pir/cpp/client.h

```diff
@@ -0,0 +1,76 @@
+//
+// Copyright 2020 the authors listed in CONTRIBUTORS.md
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef PIR_CLIENT_H_
+#define PIR_CLIENT_H_
+
+#include <string>
+
+#include "context.h"
+#include "payload.h"
+#include "util/statusor.h"
+
+namespace pir {
+
+using ::private_join_and_compute::StatusOr;
+
+class PIRClient {
+ public:
+  /**
+   * Creates and returns a new client instance, from existing parameters
+   * @param[in] params PIR parameters
+   * @returns InvalidArgument if the parameters cannot be loaded
+   **/
+  static StatusOr<std::unique_ptr<PIRClient>> Create(
+      std::shared_ptr<PIRParameters> params);
+  /**
+   * Creates a new payload request
+   * @param[in] desiredIndex Expected database value from an index
+   * @param[in] dbSize Database size
+   * @returns InvalidArgument if the index is invalid or if the encryption fails
+   **/
+  StatusOr<std::string> CreateRequest(std::size_t /*index*/) const;
```

Comment:
Shouldn't this produce a Payload, which can then be easily serialized? That way, if a test or something doesn't want to serialize, it can pass it directly to the server. i.e. decouples the serialization from the creation of the request.

Reply:
I think it's easier to pass strings, this way only the PIRPayload has to know the internal layout. We can easily change the internal layout.
Otherwise, there should be a serializer that can access the internals of the PIRPayload.

The string parameters make it easier to port it to another languages too.